app/presenters/licences/view-licence.presenter.js

```diff
@@ -15,22 +15,46 @@ const { formatLongDate } = require('../base.presenter.js')
  * @returns {Object} The data formatted for the view template
  */
 function go (licence) {
-  const { expiredDate, id, licenceRef, region, startDate } = licence
+  const { expiredDate, id, lapsedDate, licenceRef, region, revokedDate, startDate } = licence
+  const warning = _generateWarningMessage(expiredDate, lapsedDate, revokedDate)
 
   return {
     id,
     endDate: _endDate(expiredDate),
     licenceRef,
     region: region.displayName,
-    startDate: formatLongDate(startDate)
+    startDate: formatLongDate(startDate),
+    warning
   }
 }
 
 /**
- * Formats the expired date of the licence as the end date for the view
+ * Compares two end dates and chooses the correct one to return
+ *
+ * @param {date} firstEndDate - The first date to compare
+ *
+ * @param {date} secondEndDate - The second date to compare
  *
  * @module ViewLicencePresenter
  */
+function _compareEndDates (firstEndDate, secondEndDate) {
+  if (firstEndDate.date.getTime() === secondEndDate.date.getTime()) {
+    if (firstEndDate.name === 'revoked') return firstEndDate
+    if (secondEndDate.name === 'revoked') return secondEndDate
+    if (firstEndDate.name === 'lapsed') return firstEndDate
+  } else if (firstEndDate.date < secondEndDate.date) {
+    return firstEndDate
+  }
+  return secondEndDate
+}
+
+/**
+  * Formats the expired date of the licence as the end date for the view
+  *
+  * @param {date} expiredDate - The expired date to be formatted if it is in the past
+  *
+  * @module ViewLicencePresenter
```

Comment:
I might be misunderstanding something because I keep flagging these `@module` tags in JSDoc comments for removal!

According to [JSDocs](https://jsdoc.app/howto-commonjs-modules)

> In most cases, your CommonJS or Node.js module should include a standalone JSDoc comment that contains a [@module tag](https://jsdoc.app/tags-module).

This is why we normally start with one of these at the top of our files. We then don't include them inside function comments.

But I might have gotten it wrong so should stop trying to remove them from your comments. 🤷 

> Also, same comment about whether you need a JSDoc comment for this private function applies here as well!